js/todoOperations.js

```diff
@@ -0,0 +1,128 @@
+function addTaskToTable(task, todos) {
+    const tableBody = document.querySelector('.todo-table__body');
+    const row = document.createElement('tr');
+    row.classList.add('todo-table__row');
+
+    if (task.completed) {
+        row.classList.add('completed');
+    }
+
+    row.appendChild(createCell(task.id, 'todo-table__cell'));
+    row.appendChild(createCell(task.todo, 'todo-table__cell'));
+    row.appendChild(createCell(task.userId, 'todo-table__cell'));
+    row.appendChild(createCell(task.completed ? 'Completed' : 'Pending', 'todo-table__cell'));
+
+    const buttonsCell = createCell('', 'todo-table__cell');
+    const buttonsContainer = document.createElement('div');
+    buttonsContainer.classList.add('todo-table__cell--buttons');
+
+    const deleteButton = createButton('Delete', 'todo-table__button todo-table__button--delete', () => handleDelete(row, task.id));
+    const doneButton = createButton(task.completed ? 'Undo' : 'Done', 'todo-table__button todo-table__button--done', () => handleDone(row, task, doneButton, todos));
+
+    buttonsContainer.append(deleteButton, doneButton);
+    buttonsCell.appendChild(buttonsContainer);
+    row.appendChild(buttonsCell);
+    tableBody.appendChild(row);
+
+    row.children[1].addEventListener('click', () => handleEdit(row, task));
+    updateTasksCount(tableBody.children.length);
+}
+
+function handleSearch() {
+    const searchInput = document.querySelector('.task-controls__search-input');
+
+    searchInput.addEventListener('input', () => {
+        const query = searchInput.value.trim().toLowerCase();
+        const filteredTodos = todos.filter(task => task.todo.toLowerCase().includes(query));
+        displayTodos(filteredTodos);
+        updateTasksCount(filteredTodos.length);
+    });
+}
+
+async function handleDelete(row, taskId) {
```

Comment:
`handleDeleteTodo`

Reply:
Moving the !isDeleted condition to the top and returning early is a good practice because:

- **Readability:** It makes the code easier to read and understand by handling error cases or early exits at the beginning.
- **Maintainability:** It reduces the nesting level of the code, making it simpler to maintain and modify.
- **Clarity:** It clearly separates the error handling logic from the main logic of the function, improving overall code clarity.

```
async function handleDelete(row, taskId) {
    if (!confirm('Are you sure you want to delete this task?')) return;

    const isDeleted = await deleteTodoFromAPI(taskId);

    if (!isDeleted) {
        alert('Failed to delete the TODO from the server.');
        return;
    }

    row.remove();

    const updatedTodos = todos.filter(todo => todo.id !== taskId);
    saveTodosToLocalStorage(updatedTodos);

    todos.length = 0;
    todos.push(...updatedTodos);

    const searchInput = document.querySelector('.task-controls__search-input');
    const query = searchInput.value.trim().toLowerCase();

    if (query) {
        const displayedTodos = updatedTodos.filter(task => task.todo.toLowerCase().includes(query));
        displayTodos(displayedTodos);
        updateTasksCount(displayedTodos.length);
    } else {
        displayTodos(updatedTodos);
        updateTasksCount(updatedTodos.length);
    }
}
```

Reply:
We can reduce the displaying functionlity as

```
const displayedTodos = updatedTodos.filter(task => 
    !query || task.todo.toLowerCase().includes(query)
);

displayTodos(displayedTodos);
updateTasksCount(displayedTodos.length);
```